dev/tools/bin/find_commit.dart

```diff
@@ -0,0 +1,93 @@
+
+import 'dart:io';
+
+const bool debugLogging = false;
+
+void log(String message) {
+  if (debugLogging)
+    print(message);
+}
+
+class Commit {
+  Commit(this.hash, this.timestamp);
+
+  final String hash;
+  final DateTime timestamp;
+
+  static String formatArgument = '--format=%H %cI';
+
+  static Commit parse(String line) {
+    final int space = line.indexOf(' ');
+    return Commit(line.substring(0, space), DateTime.parse(line.substring(space+1, line.length).trimRight()));
+  }
+
+  static List<Commit> parseList(String lines) {
+    return lines.split('\n').where((String line) => line.isNotEmpty).map(parse).toList().reversed.toList();
+  }
+}
+
+String findCommit({
+  required String primaryRepoDirectory,
+  required String primaryBranch,
+  required String primaryTrunk,
+  required String secondaryRepoDirectory,
+  required String secondaryBranch,
+}) {
+  final Commit anchor;
+  if (primaryBranch == primaryTrunk) {
+    log('on $primaryTrunk, using last commit as anchor');
+    anchor = Commit.parse(git(primaryRepoDirectory, <String>['log', Commit.formatArgument, '--max-count=1', primaryBranch]));
+  } else {
+    final List<Commit> branchCommits = Commit.parseList(git(primaryRepoDirectory, <String>['log', Commit.formatArgument, primaryBranch]));
+    final List<Commit> trunkCommits = Commit.parseList(git(primaryRepoDirectory, <String>['log', Commit.formatArgument, primaryTrunk]));
+    if (branchCommits.isEmpty || trunkCommits.isEmpty || branchCommits.first.hash != trunkCommits.first.hash)
+      throw StateError('Branch $primaryBranch does not seem to have a common history with trunk $primaryTrunk.');
+    if (branchCommits.last.hash == trunkCommits.last.hash) {
+      log('$primaryBranch is even with $primaryTrunk, using last commit as anchor');
+      anchor = trunkCommits.last;
+    } else {
+      int index = 0;
+      while (branchCommits.length > index && trunkCommits.length > index && branchCommits[index].hash == trunkCommits[index].hash)
+        index += 1;
+      log('$primaryBranch branched from $primaryTrunk ${branchCommits.length - index} commits ago, trunk has advanced by ${trunkCommits.length - index} commits since then.');
+      anchor = trunkCommits[index - 1];
+    }
+  }
+  return git(secondaryRepoDirectory, <String>[
+    'log',
+    '--format=%H',
+    '--until=${anchor.timestamp.toIso8601String()}',
+    '--max-count=1',
+    secondaryBranch,
+  ]);
+}
+
+String git(String workingDirectory, List<String> arguments) {
+  final ProcessResult result = Process.runSync('git', arguments, workingDirectory: workingDirectory);
+  if (result.exitCode != 0 || '${result.stderr}'.isNotEmpty)
+    throw ProcessException('git', arguments, '${result.stdout}${result.stderr}', result.exitCode);
+  return '${result.stdout}';
+}
+
+void main(List<String> arguments) {
+  if (arguments.length != 1) {
+    print('This script should be run with a single argument, the path to the secondary repository.');
```

Comment:
can this take two arguments, so that the primary repo must be provided also?

Reply:
I considered that (actually my first version of this literally took `git:...` URLs and cloned the repos in `/tmp` so it was completely isolated) but it seemed easier to just run it from inside the main repo, which presumably is where the calling script is likely to be sitting anyway.

That said I'm happy to change this to be whatever is useful for you. What API do you think is most useful?

Reply:
in LUCI, it is easier to pass paths as absolute then to have to change directory for a specific step of a build.

Reply:
Aren't we moving away from LUCI limitations? I thought the idea was to not have to worry about LUCI configuration issues and recipes and such and just describe shards in the repo itself, at which point presumably you just invoke a bash script or some such.

Reply:
We still are dependent on LUCI for the file system setup and the current working directory.

Reply:
The starting working directory sure, but can't the script then just change directories to wherever?

Anyway that's probably a discussion for another forum, if you want a second argument I'm happy to provide it.